### PR TITLE
k3d: fetch K3s version from channel server

### PR DESCRIPTION
**Pull request by iwilltry42**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

File: Formula/k3d.rb [hunk 1]
```diff
@@ -19,14 +20,17 @@ class K3d < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "e117dd2b6ebf56a0def38e790920f5c9ac9bbdbfe1abb617c933f46db8770f40"
   end
 
+  depends_on "gnu-sed" => :build
   depends_on "go" => :build
 
+  # curl and sed used to fetch/parse latest stable K3s version
+  uses_from_macos "curl" => :build
```
Comment (carlocab) [inline Formula/k3d.rb:27]: `curl` is always installed (Homebrew doesn't work if you don't have a pre-existing `curl`), so unless you need Homebrew `curl` on Linux specifically, you don't need to declare this.

File: Formula/k3d.rb [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
-  sha256 "85fff692fd3180fb9d050a0e25a99718cde7539e53ef0ffcb433fd19dd56061f"
+  url "https://github.com/rancher/k3d.git",
+    tag:      "v5.0.3",
+    revision: "89d979287e0cff56ac8e87e5e0ecab9dc3b6ba87"
   license "MIT"
 
   livecheck do
     url :stable
-    regex(/^v?(\d+(?:\.\d+)+)$/i)
+    strategy :github_latest
```
Comment (carlocab) [inline Formula/k3d.rb:11]: Why is this needed?

File: Formula/k3d.rb [hunk 3]
```diff
@@ -42,7 +46,7 @@ def install
   end
 
   test do
-    assert_match "k3d version v#{version}\nk3s version latest (default)", shell_output("#{bin}/k3d --version")
+    assert_match "k3d version v#{version}", shell_output("#{bin}/k3d version | head -1")
```
Comment (carlocab) [inline Formula/k3d.rb:49]:
```suggestion
    assert_match "k3d version v#{version}", shell_output("#{bin}/k3d version")
```
`assert_match` only matches substrings, so you don't need to pipe to `head`. If you really only want the first line, you can use `.split.first`. (`split` might needs to take a `"\n"` argument; I forget if this splits new lines by default.)

File: Formula/k3d.rb [hunk 4]
```diff
@@ -19,14 +20,17 @@ class K3d < Formula
...
+  uses_from_macos "curl" => :build
+
   def install
-    system "go", "build",
-           "-mod", "vendor",
-           "-ldflags", "-s -w -X github.com/rancher/k3d/v#{version.major}/version.Version=v#{version}"\
-                       " -X github.com/rancher/k3d/v#{version.major}/version.K3sVersion=latest",
-           "-trimpath", "-o", bin/"k3d"
+    # Build k3d
+    system  "make", "build", # Makefile includes fetching the latest stable K3s version
```
Comment (carlocab) [inline Formula/k3d.rb:31]: What's wrong with the current way of building this?

File: Formula/k3d.rb (hunk 3 above)
Comment (iwilltry42) [reply Formula/k3d.rb:49]: Good to know, thanks!

File: Formula/k3d.rb (hunk 2 above)
Comment (iwilltry42) [reply Formula/k3d.rb:11]: It made sense to me after checking the docs when I saw it in other formulae, as per https://rubydoc.brew.sh/Homebrew/Livecheck/Strategy/GithubLatest.html (and the regex is the default in that strategy).

File: Formula/k3d.rb (hunk 1 above)
Comment (iwilltry42) [reply Formula/k3d.rb:27]: Good point, I'll remove this then.. was going to argue about it, but if curl is a brew requirement anyway, it doesn't make sense for real (it's in a few other formulae as well).

File: Formula/k3d.rb (hunk 4 above)
Comment (iwilltry42) [reply Formula/k3d.rb:31]: Two things:
1. It's effectively just the exact same commands that are already included in the Makefile (why rewrite it here and having to change it in two places in case anything changes? :man_shrugging: )
2. The Makefile fills the K3s version variable by doing some curl + sed magic (would require even more lines in the formula otherwise)

File: Formula/k3d.rb (hunk 1 above)
Comment (iwilltry42) [reply Formula/k3d.rb:27]: :heavy_check_mark:

File: Formula/k3d.rb (hunk 3 above)
Comment (iwilltry42) [reply Formula/k3d.rb:49]: :heavy_check_mark:

**Comment by iwilltry42** [comment]
@carlocab , thanks for your feedback. I force-pushed your proposed changes to keep the commit history clean :+1:

File: Formula/k3d.rb (hunk 2 above)
Comment (carlocab) [reply Formula/k3d.rb:11]: Does the current livecheck produce wrong results? If it doesn't, let's keep it as it is. `:github_latest` gets rate limited so we only use it when we actually have evidence that it's needed.

File: Formula/k3d.rb (hunk 4 above)
Comment (carlocab) [reply Formula/k3d.rb:29]: Here are a few reasons why:
1. Git clones are slow and expensive, and it really sucks to have to use them just to fetch version information.
2. The build is more transparent when it's not hidden behind a `Makefile`. That's useful for users reading the formula who just want to understand what goes on when something gets built. (It's not always possible, since some `Makefiles` are too complicated, but Go builds are usually simple enough it's not too bad to do it all in the formula.)
3. It's probably pretty useful to have the version information generated from a Homebrew installation be different from other installation methods. That is, unless our builds are byte-for-byte identical to other installations, but I doubt they are (or that they'll continue to be if they are at the moment). This simplifies tracking down bugs, for example.

Comment (SMillerDev) [reply Formula/k3d.rb:29]: > It's effectively just the exact same commands that are already included in the Makefile (why rewrite it here and having to change it in two places in case anything changes? 🤷‍♂️ )

Because we can optimize this more than we can a Makefile.

> The Makefile fills the K3s version variable by doing some curl + sed magic (would require even more lines in the formula otherwise)

Curl + sed magic does not sound very stable or reliable, while our  variables are.

File: Formula/k3d.rb (hunk 2 above)
Comment (iwilltry42) [reply Formula/k3d.rb:11]: Ah if it's rate-limited then I'll change it back.
As most other changes it was inspired by e.g. the k3sup and k9s formula.
But this makes sense 👍

File: Formula/k3d.rb (hunk 1 above)
Comment (carlocab) [reply Formula/k3d.rb]: Other formulae use it because they require a new version of `curl` (i.e. the minimum required by Homebrew doesn't suffice). At least, that's the idea. There are probably some that we can get rid of.

File: Formula/k3d.rb (hunk 4 above)
Comment (iwilltry42) [reply Formula/k3d.rb:29]: Okay, so no problem for me to keep the go build logic in here if you think it's the better way to go and I don't think we'll have many more variable injections in the future.

With regards to the curl + sed magic: were just fetching a file from the K3s channel server and parse it to get the latest tag of the stable channel so we can inject the K3s version variable (latest is not good here).
So it's not about the k3d version (which we use from homebrew), but about K3s which is used by k3d, so we have to use curl+sed (or a more "stable" way with jq, but that's not installed by default on many systems (or some ruby if that's possible)).

Comment (carlocab) [reply Formula/k3d.rb:29]: Ruby can parse JSON natively, so it should be a lot easier than with `sed`.

Comment (Bo98) [reply Formula/k3d.rb:29]: We're slowly requiring `std_go_args` to be used for all Go dependents. It should have been used here before.

Makefiles are fine in general, but they typically make it impossible to use our `std_go_args` or make it complicated enough that it's not really any simpler using the Makefile.

Looking at the Makefile here for example, it looks like it is possible to override but not append to LDFLAGS, so we would have to handle the versioning logic ourselves anyway.

Comment (iwilltry42) [reply Formula/k3d.rb]: Alright, then I'll go with the new standards here :+1:

**Comment by iwilltry42** [comment]
Hey there :wave: I've updated the PR to use Ruby code for fetching the K3s version from the channel server and to use the proposed ways for livechecking, variable usage and pulling the release.
I'm not super familiar with Ruby, so there may be quite some room for enhancement, but at least it works :grimacing:

File: Formula/k3d.rb [hunk 5]
```diff
@@ -22,31 +25,34 @@ class K3d < Formula
...
 
     # Install fish completion
-    output = Utils.safe_popen_read("#{bin}/k3d", "completion", "fish")
+    output = Utils.safe_popen_read(bin/name, "completion", "fish")
     (fish_completion/"k3d.fish").write output
   end
 
   test do
-    assert_match "k3d version v#{version}\nk3s version latest (default)", shell_output("#{bin}/k3d --version")
+    assert_match "k3d version v#{version}\nk3s version latest (default)", shell_output("#{bin/name} --version")
```
Comment (carlocab) [inline Formula/k3d.rb:51]: Looks like you need to update the assertion here, because it no longer says `version latest`.

File: Formula/k3d.rb [hunk 6]
```diff
@@ -1,3 +1,6 @@
+require "net/http"
+require "json"
```
Comment (carlocab) [inline Formula/k3d.rb:2]:
```suggestion
```
This isn't needed.

File: Formula/k3d.rb [hunk 7]
```diff
@@ -1,3 +1,6 @@
+require "net/http"
```
Comment (carlocab) [inline Formula/k3d.rb:1]: Can we do this just inside the `install` method?

File: Formula/k3d.rb (hunk 5 above)
Comment (iwilltry42) [reply Formula/k3d.rb:51]: True, that got reverted with the rest of the original changes in this PR :heavy_check_mark:

File: Formula/k3d.rb (hunk 6 above)
Comment (iwilltry42) [reply Formula/k3d.rb:2]: :check

File: Formula/k3d.rb (hunk 7 above)
Comment (iwilltry42) [reply Formula/k3d.rb:1]: :heavy_check_mark:

File: Formula/k3d.rb [hunk 8]
```diff
@@ -22,31 +22,35 @@ class K3d < Formula
...
     system "go", "build",
            "-mod", "vendor",
-           "-ldflags", "-s -w -X github.com/rancher/k3d/v#{version.major}/version.Version=v#{version}"\
-                       " -X github.com/rancher/k3d/v#{version.major}/version.K3sVersion=latest",
-           "-trimpath", "-o", bin/"k3d"
+           *std_go_args(ldflags: "-s -w -X github.com/rancher/k3d/v#{version.major}/version.Version=v#{version} -X github.com/rancher/k3d/v#{version.major}/version.K3sVersion=#{k3s_version}")
 
     # Install bash completion
-    output = Utils.safe_popen_read("#{bin}/k3d", "completion", "bash")
+    output = Utils.safe_popen_read(bin/name, "completion", "bash")
```
Comment (SMillerDev) [inline Formula/k3d.rb:36]: Can we just use the more verbose
```suggestion
    output = Utils.safe_popen_read(bin/"k3d", "completion", "bash")
```

File: Formula/k3d.rb [hunk 9]
```diff
@@ -22,31 +22,35 @@ class K3d < Formula
...
+    resp = Net::HTTP.get(uri)
+    resp_json = JSON.parse(resp)
+    k3s_version = resp_json["data"].find { |channel| channel["id"]=="stable" }["latest"].sub("+", "-")
+
     system "go", "build",
            "-mod", "vendor",
-           "-ldflags", "-s -w -X github.com/rancher/k3d/v#{version.major}/version.Version=v#{version}"\
-                       " -X github.com/rancher/k3d/v#{version.major}/version.K3sVersion=latest",
-           "-trimpath", "-o", bin/"k3d"
+           *std_go_args(ldflags: "-s -w -X github.com/rancher/k3d/v#{version.major}/version.Version=v#{version} -X github.com/rancher/k3d/v#{version.major}/version.K3sVersion=#{k3s_version}")
```
Comment (SMillerDev) [inline Formula/k3d.rb:33]: Can you define the ldflags in a variable above and just reference that instead? This is harder to read

File: Formula/k3d.rb (hunk 8 above)
Comment (iwilltry42) [reply Formula/k3d.rb]: :heavy_check_mark:

File: Formula/k3d.rb (hunk 9 above)
Comment (iwilltry42) [reply Formula/k3d.rb]: :heavy_check_mark:

File: Formula/k3d.rb (hunk 8 above)
Comment (carlocab) [reply Formula/k3d.rb]: I do think `name` is pretty appropriate here so that `brew extract` just works without any tweaks given that we're using `std_go_args`.

Comment (iwilltry42) [reply Formula/k3d.rb]: @carlocab , is there any action needed still or can this be merged? :)

**Review by carlocab — DISMISSED** [review]

**Comment by SMillerDev** [comment]
Rebased the PR so it can be merged

**Review by carlocab — APPROVED** [review]
